sklearn/cluster/kernel_kmeans_.py

```diff
@@ -0,0 +1,205 @@
+        The type of kernel to be used in kernel k means algorithm.
+
+    gamma : float, optional
+
+    degree : int, default: 3
+
+    coef0 : int, default: 1
+
+    kernel_params : optional
+        Additional kernel parameters.
+
+    verbose : int, default: 0
+        Verbosity mode.
+
+    Attributes
+    ----------
+
+    sample_weight_ : array-like, shape=(n_samples,)
+
+    labels_ : shape=(n_samples,)
+        Labels of each point.
+
+    within_distances_ : array, shape=(n_clusters,)
+        Distance update.
+
+    X_fit_ : array-like, shape=(n_samples, n_features)
+        It is the data used in clustering.
+    """
+
+    def __init__(self, n_clusters=3, max_iter=300, tol=1e-3, random_state=None,
+                 kernel="linear", gamma=None, degree=3, coef0=1,
+                 kernel_params=None, verbose=0):
+        self.n_clusters = n_clusters
+        self.max_iter = max_iter
+        self.tol = tol
+        self.random_state = random_state
+        self.kernel = kernel
+        self.gamma = gamma
+        self.degree = degree
+        self.coef0 = coef0
+        self.kernel_params = kernel_params
+        self.verbose = verbose
+
+    @property
+    def _pairwise(self):
+        return self.kernel == "precomputed"
+
+    def _get_kernel(self, X, Y=None):
+        if callable(self.kernel):
+            params = self.kernel_params or {}
+        else:
+            params = {"gamma": self.gamma,
+                      "degree": self.degree,
+                      "coef0": self.coef0}
+        return pairwise_kernels(X, Y, metric=self.kernel,
+                                filter_params=True, **params)
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Compute k-means clustering.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+
+        sample_weight : array-like, shape=(n_samples,)
+        """
+        n_samples = X.shape[0]
+
+        K = self._get_kernel(X)
+
+        sw = sample_weight if sample_weight else np.ones(n_samples)
+        self.sample_weight_ = sw
+
+        rs = check_random_state(self.random_state)
+        self.labels_ = rs.randint(self.n_clusters, size=n_samples)
+
+        dist = np.zeros((n_samples, self.n_clusters))
+        self.within_distances_ = np.zeros(self.n_clusters)
+
+        for it in xrange(self.max_iter):
+            dist.fill(0)
+            self._compute_dist(K, dist, self.within_distances_,
+                               update_within=True)
+            labels_old = self.labels_
+            self.labels_ = dist.argmin(axis=1)
+
+            # Compute the number of samples whose cluster did not change
+            # since last iteration.
+            n_same = np.sum((self.labels_ - labels_old) == 0)
+            if 1 - float(n_same) / n_samples < self.tol:
+                if self.verbose:
+                    print "Converged at iteration", it + 1
+                break
+
```

Comment:
I believe one extra label update is needed here: because cluster centers have been updated but cluster labels haven't, it's possible for the loop to finish in a way that ``self.fit(X).predict(X)`` returns different results than ``self.fit_predict(X)``.